vinyl: set bloom fpr automatically

EmirVildanov opened the pull request:
This pull request is dedicated to implementing the optimized bloom_fpr calculation algorithm.
It consists of two parts:
- Changes in core logic of bloom_fpr calculation
- Benchmark in /perf subfolder for analyzing patch impact

EmirVildanov committed:
vinyl: add bloom_fpr optimal calculation

@TarantoolBot document
Title: optimized bloom_fpr calculation

The optimal bloom_fpr is calculated
using transformed optimization from Monkey:
https://stratos.seas.harvard.edu/files/stratos/files/monkeykeyvaluestore.pdf
article: the calculation based not on a run's level
in LSM tree, but on it's size.

vy_scheduler.c added functions:
vy_task_find_range_bloom_fpr(lsm, range,
new_run_stmts_count, W, total_stmts_count, skip_n)
Calculate optimal bloom_fpr for given range.
Works for runs created during compaction and as
a subfunction during dump run creation.

Parameters:
* lsm — LSM-tree we are working with
* range — concrete range we are working with
* new_run_stmts_count — statements count in
the newly created after dump/compaction run
* W — total number of runs in the range after
operation (dump/compaction)
* total_stmts_count — total number of statements
in the range after operation (dump/compaction)
* skip_n — number of runs to skip during the range
traverse as they are compacted into one new run

Returns:
optimally calculated bloom_fpr

vy_task_find_lsm_bloom_fpr(lsm, new_run_stmts_count)
Calculate optimal bloom_fpr for given lsm.
Using vy_task_find_range_bloom_fpr inside.

Parameters:
* lsm — LSM-tree we are working with
* new_run_stmts_count — statements count in
the newly created after dump run

Returns:
optimally calculated bloom_fpr for dump

See also #3969
NO_TEST=internal

CuriousGeorgiy commented on perf/bloom_filter/results.yml:
```diff
@@ -0,0 +1,355 @@
+builds:
```
Please drop this file, artifcats should not be in VCS.

EmirVildanov replied:
Dropped the file. Will post results to the commit message as soon as benchmark is fixed.

CuriousGeorgiy commented on src/lib/salad/heap.h:
```diff
@@ -389,7 +389,7 @@ HEAP(reserve)(heap_t *heap)
 	if (harr == NULL)
 		return -1;
-	heap->harr = harr;
+	heap->harr = (struct heap_node **)harr;
```
Please drop this change, seems unrelated to the patch.

EmirVildanov replied:
As soon as there is a compiler link sequence of box.cc -> vy_lsm.h -> vy_range.h -> salad/heap.h and C++ is on duty there occurs a compiler error: invalid conversion from ‘void*’ to ‘heap_node**’ [-fpermissive].
Can't drop the change.

CuriousGeorgiy commented on src/box/vy_stmt.h:
```diff
@@ -257,7 +257,7 @@ vy_stmt_set_n_upserts(struct tuple *stmt, uint8_t n)
 vy_stmt_is_key_format(const struct tuple_format *format)
 {
-	struct vy_stmt_env *env = format->engine;
+	struct vy_stmt_env *env = (struct vy_stmt_env *)format->engine;
```
Please drop this change, seems unrelated to the patch.

EmirVildanov replied:
See above^^^. Resolved.

CuriousGeorgiy commented on src/box/box.h:
```diff
@@ -558,6 +558,19 @@ box_sequence_reset(uint32_t seq_id);
+ * \retval 0 on success
+ */
+API_EXPORT int
+box_bloom_bsize(uint32_t space_id, uint32_t index_id, size_t *result);
```
I would rather name it `box_index_bloom_bsize`.

EmirVildanov replied:
Done.

CuriousGeorgiy commented on perf/utils/timer.h:
```diff
@@ -0,0 +1,28 @@
+/** Primitive wrapper to measure execution time. */
+class PerfTimer {
+public:
+	PerfTimer() {};
```
Looks redundant.

EmirVildanov replied:
Removed it.

CuriousGeorgiy commented on perf/utils/randb.h:
```diff
@@ -0,0 +1,11 @@
+#include <random>
+
+struct RandomByteGenerator {
+	RandomByteGenerator(unsigned min = 0, unsigned max = UCHAR_MAX) :
```
Why is it `UCHAR_MAX`?

EmirVildanov replied:
Deleted randb.h file as it useless for this benchmark.

CuriousGeorgiy commented on src/box/box.cc:
```diff
@@ -2834,6 +2835,28 @@ box_session_push(const char *data, const char *data_end)
+	index = index_find(space, index_id);
+	if (index == NULL)
+		return -1;
+	struct vy_lsm *lsm = vy_lsm(index);
```
**memtx* spaces not handled.

EmirVildanov replied:
How can I check that space I get from `space_cache_find` is a **memtx** space?

EmirVildanov replied:
Should I return -1 in case of memtx engine found?

CuriousGeorgiy replied:
Yep.

EmirVildanov replied:
Done.

CuriousGeorgiy commented on src/box/box.cc:
```diff
@@ -2834,6 +2835,28 @@ box_session_push(const char *data, const char *data_end)
+	struct vy_lsm *lsm = vy_lsm(index);
+	ssize_t bsize = lsm->bloom_size;
+	if (bsize < 0)
+		return -1;
```
This logic should be moved to `vy_lsm.{h, c}`

EmirVildanov replied:
Am I right that I should move only the logic of getting bloom_size field from lsm so it looks like `ssize_t bsize = vy_lsm_bloom_size(lsm)`?

EmirVildanov replied:
1.) Moved index to lsm casting in vy_lsm_bloom_size()
2.) Shouldn't we return -1 in case of en error?

CuriousGeorgiy replied:
No, I mean how can bloom_size be possibly negative?

EmirVildanov replied:
For some reason I thought that returned `bloom_size` has `ssize_t` type and decided to check if it's equals to -1.
Deleted if statement.

CuriousGeorgiy commented on perf/bloom_filter/launch.lua.in:
```diff
@@ -0,0 +1,105 @@
+local bench_stop_func = "bench.stop"
+local bench_funcs = { bench_run_func, bench_init_func, bench_stop_func }
+local primary_index_name = "pk"
+-- naming is followed in order to Lua pairs() call iterate through them correct
```
I didn't get the idea...

EmirVildanov replied:
As soon as Lua `pairs` function does not iterate through `spaces_lookup_cost_coeffs_pairs` table values in the order they were declared, naming convention of view like "read<Number><Name>" at first seemed good for solving the order issue.
Removed comment with strange spaces naming.

EmirVildanov replied:
Update: decided to leave the naming as it's more comfortable to analyze such names while looking at benchmark results (they are logically sorted in lexicographic order).

CuriousGeorgiy commented on perf/bloom_filter/launch.lua.in:
```diff
@@ -0,0 +1,105 @@
+
+local function schema_create()
+    -- made range big enough so we can benefit from working on different levels
+    local range_size_value = 134217728 * 24
```
What is the first magic number?

EmirVildanov replied:
I don't remember why I have chosen it.
I wanted to use default value for `vinyl_range_size` that is 1073741824 for Tarantool < 1.10.2. Changed it.

CuriousGeorgiy commented on perf/bloom_filter/launch.lua.in:
```diff
@@ -0,0 +1,105 @@
+        box.space[space_name]:format({
+            {name = 'number', type = 'unsigned'},
+            {name = 'name', type = 'string'},
+        })
```
Seems redundant to me.

EmirVildanov replied:
Removed `format` function call.

CuriousGeorgiy commented on perf/bloom_filter/launch.lua.in:
```diff
@@ -0,0 +1,105 @@
+
+    box.cfg{
+        -- gh-2278 vinyl does not support DDL/DML if wal_mode = 'none'
+        -- wal_mode = 'none',
```
Just drop these lines please.

EmirVildanov replied:
Done.

CuriousGeorgiy commented on src/box/vy_lsm.h:
```diff
@@ -176,7 +176,7 @@ vy_lsm_env_destroy(struct vy_lsm_env *env);
  *   key, *augmented* with the parts of the primary key from the
```
These are actually two distinct changes: you either have to split this into two commits for `box` and `salad` components, or just drop these changes.

EmirVildanov replied:
Dropped the commit.

CuriousGeorgiy commented on src/box/alter.cc:
```diff
@@ -244,6 +244,12 @@ index_opts_decode(struct index_opts *opts, const char *map,
 			 "less than or equal to 1");
 		return -1;
 	}
+	if (opts->lookup_cost_coeff <= 0 || opts->lookup_cost_coeff > 1) {
```
Why is it [0; 1]? AFAIC, it can be any positive integer.

EmirVildanov replied:
Discussed it.
The motivation of using [0; 1] instead of [0; MAX_R_VALUE] is that:
- MAX_R_VALUE is calculated dynamically from the number of runs in the range
- It will be more comfortable for user to choose value in [0;1] range

CuriousGeorgiy commented on src/box/box.cc:
```diff
@@ -1270,6 +1271,10 @@ box_check_vinyl_options(void)
 		tnt_raise(ClientError, ER_CFG, "vinyl_bloom_fpr",
 			  "must be greater than 0 and less than or equal to 1");
 	}
+	if (lookup_cost_coeff <= 0 || lookup_cost_coeff > 1) {
```
Ditto.

EmirVildanov replied:
Resolved ^^^.

CuriousGeorgiy commented on src/box/index_def.c:
```diff
@@ -48,6 +48,7 @@ const struct index_opts index_opts_default = {
 	/* .run_count_per_level = */ 2,
 	/* .run_size_ratio      = */ 3.5,
 	/* .bloom_fpr           = */ 0.05,
+    /* .lookup_cost_coeff   = */ 0.5,
```
This requires discussion.

EmirVildanov replied:
Fixed the indent.
Seems like we should get default value by experimenting with the select queries latency so it approximately equals latency with bloom_fpr = 0.05.

CuriousGeorgiy replied:
Looks like you didn't fix indent.

EmirVildanov replied:
Fixed it (Had to rebase and move it from one commit to another).

CuriousGeorgiy commented on src/box/index_def.h:
```diff
@@ -151,8 +151,16 @@ struct index_opts {
+    /* Coefficient of worst-case read scenario:
+     * the lower parameter, the higher read performance
+     * with more memory consumption.
+     */
```
Please elaborate on the comment — it should rather be 'mean amount of disk lookups'.

EmirVildanov replied:
Should I still change the comment as we discussed that `lookup_cost_coeff` is not a mean amount of lookups?

CuriousGeorgiy replied:
Yes.

EmirVildanov replied:
Done.

CuriousGeorgiy commented on src/box/lua/load_cfg.lua:
```diff
@@ -64,6 +64,7 @@ local default_cfg = {
     vinyl_range_size          = nil, -- set automatically
     vinyl_page_size           = 8 * 1024,
     vinyl_bloom_fpr           = 0.05,
+    vinyl_lookup_cost_coeff   = 0.5,
```
Ditto.

CuriousGeorgiy commented on src/box/vy_range.c:
```diff
@@ -345,6 +345,7 @@ vy_range_update_compaction_priority(struct vy_range *range,
 	slice = rlist_last_entry(&range->slices, struct vy_slice, in_range);
 	size = MAX(slice->count.bytes, 1);
 	slice = rlist_first_entry(&range->slices, struct vy_slice, in_range);
+    /** Should we change it to while cycle? */
```
Please drop this comment.

EmirVildanov replied:
Done.

CuriousGeorgiy commented on src/box/vy_scheduler.c:
```diff
@@ -1402,7 +1533,12 @@ vy_task_dump_new(struct vy_scheduler *scheduler, struct vy_worker *worker,
+	double bloom_fpr = vy_task_find_lsm_bloom_fpr(
+					lsm,
+					dump_stmts_count);
+	if (bloom_fpr <= 0 || bloom_fpr > 1)
```
This shouldn't be possible.

EmirVildanov replied:
Fixed it.

CuriousGeorgiy commented on src/box/vy_scheduler.c:
```diff
@@ -1709,7 +1854,14 @@ vy_task_compaction_new(struct vy_scheduler *scheduler, struct vy_worker *worker,
+			range->slice_count - range->compaction_priority + 1,
+			range->count.rows,
+			range->compaction_priority);
+	if (bloom_fpr <= 0 || bloom_fpr > 1)
```
This shouldn't be possible.

EmirVildanov replied:
Done ^^^.

CuriousGeorgiy commented on src/box/vy_scheduler.c:
```diff
@@ -1116,6 +1117,131 @@ vy_task_write_run(struct vy_task *task, bool no_compression)
+ * @return optimally calculated bloom_fpr for dump
+ */
+double
+vy_task_find_lsm_bloom_fpr(
```
I don't understand the name of this function.

EmirVildanov replied:
I would like to call it `vy_task_find_bloom_fpr_for_run_in_range`, but the name seems too long for me. Can't come up with better name.
As a variant I can move this function to vy_run.c file and name it like `vy_run_find_bloom_fpr_in_range`, but I don't think that this logic should lie in vy_run.c file.

CuriousGeorgiy replied:
Let's move it to `box/vy_stmt.{h, c}` and call it `vy_bloom_optimal_fpr_for_dump`.

EmirVildanov replied:
I think that vy_stmt.h may be not the best location because adding vy_lsm.h header in it causes an error `implicit declaration of function ‘vy_entry_compare’`. Static function `vy_entry_compare` used in vy_stmt.h -> vy_lsm.h -> vy_cache.h before declaration.

CuriousGeorgiy commented on src/box/vy_scheduler.c:
```diff
@@ -1116,6 +1117,131 @@ vy_task_write_run(struct vy_task *task, bool no_compression)
+ * @return optimally calculated bloom_fpr
+ */
+double
+vy_task_find_range_bloom_fpr(
```
I don't understand the name of this function.

CuriousGeorgiy replied:
Let's move it to `box/vy_stmt.{h, c}` and call it `vy_bloom_optimal_fpr`.

EmirVildanov replied:
See problem above ^^^.

EmirVildanov committed:
vinyl: added box API
function for getting bloom
size memory used by space

NO_CHANGELOG=internal
NO_DOC=internal
NO_TEST=internal

EmirVildanov committed:
perf: added benchmark for
bloom filter

Benchmark consists of two parts:
* Lua (for configuration)
* C (for clean benchmarking)

Here's the settings you may change in Lua config:
* count of select queries (inserted tuples count)
* spaces lookup_cost_coeff values
* name of the benchmarking build (e.g. "Master")

Before each test we prepare vector of tuples
of size tuple_count and then make the same
number of select queries, recording query time
and bloom_size values.

Here's the results obtained on my PC (Intel Core i5-8300H 8 X 2.30GHz)
for tuple_count = 2^18 for each space:
Space_name------Master_select(QPS)---Changed_select(QPS)
read1VeryFast---17k------------------19k
read2Fast-------16k------------------19k
read3Middle-----16k------------------19k
read4Slow-------17k------------------18k
read5VerySlow---16k------------------10k

Space_name------Master_bloom_size(bytes)---Changed_bloom_size(bytes)
read1VeryFast---235k-----------------------240k
read2Fast-------232k-----------------------190k
read3Middle-----232k-----------------------70k
read4Slow-------233k-----------------------64k
read5VerySlow---233k-----------------------0k

Co-authored-by: Nikita
Pettik
<korablev@tarantool.org>

NO_CHANGELOG=internal
NO_DOC=internal
NO_TEST=internal

EmirVildanov committed:
vinyl: added tests for optimal bloom_fpr calculation

NO_CHANGELOG=internal
NO_DOC=internal

EmirVildanov commented on perf/bloom_filter/launch.lua.in:
```diff
@@ -0,0 +1,105 @@
+    box.func[bench_init_func]:call({"@BLOOM_FILTER_BUILD_NAME@"})
+    for space_name, lookup_coeff in pairs(spaces_lookup_cost_coeffs_pairs) do
+        box.func[bench_run_func]:call({ space_name, box.space[space_name].id })
+        -- print(box.space[space_name].index[primary_index_name]:stat().disk.bloom_size)
```
Dropped this line.

CuriousGeorgiy commented on test/sql/vinyl-opts-cfg.lua:
```diff
@@ -5,6 +5,7 @@
 --
 box.cfg {
     vinyl_bloom_fpr = 0.1,
+    vinyl_lookup_cost_coeff = 0.5,
```
Why do you set the default value here? I guess it needs to be either dropped or set correspondingly to `vinyl_bloom_fpr` here (e.g., two times greater than the default value).

EmirVildanov replied:
The question is what default value of `lookup_cost_coeff` we should use. I think we should take it from benchmark results: with default value of `lookup_cost_coeff` diff between master version (where `bloom_fpr` always equal 0.05) and changed version RQS must be minimal.

CuriousGeorgiy commented on src/box/vy_scheduler.c:
```diff
@@ -1116,6 +1117,140 @@ vy_task_write_run(struct vy_task *task, bool no_compression)
+	if (new_run_stmts_count == total_stmts_count) {
+		/** By definition of R. May also look at formulas ahead.
+		 *  R <= 1 in this case as soon as W = 1.
+		 *  It also means that skip_n == range->slice_count.
```
Please add an assertion for this.

Also, please add assertions everywhere for the returned value's range.

EmirVildanov replied:
Why assertion is better than handling such case as an exception in function like `vy_task_compaction_new` using goto?

CuriousGeorgiy commented on src/box/vy_scheduler.c:
```diff
@@ -1116,6 +1117,140 @@ vy_task_write_run(struct vy_task *task, bool no_compression)
+
+	if (new_run_stmts_count > last_slice_stmts_count) {
+		/** The only case is when the size of dumped run
+		 *  is bigger than the size of the last run.
```
Please reflect this in the comment to `index_opts.bloom_fpr`.

EmirVildanov replied:
Done.

CuriousGeorgiy commented on src/box/vy_scheduler.c:
```diff
@@ -1116,6 +1117,140 @@ vy_task_write_run(struct vy_task *task, bool no_compression)
+ * @param range
+ * @param new_run_stmts_count statements count in the newly created after
+ * dump/compaction run
+ * @param W total number of runs in the range
```
Let's name the parameter properly and create an alias for it locally.

EmirVildanov replied:
Done.

CuriousGeorgiy commented on src/box/vy_scheduler.c:
```diff
@@ -1116,6 +1117,140 @@ vy_task_write_run(struct vy_task *task, bool no_compression)
+
+	if (new_run_stmts_count == total_stmts_count) {
+		/** By definition of R. May also look at formulas ahead.
+		 *  R <= 1 in this case as soon as W = 1.
```
Please add an assertion for this.

EmirVildanov replied:
1.) Done.
2.) Why assertion is better than handling such case as an exception in function like `vy_task_compaction_new` using goto?

EmirVildanov replied:
Get the answer.

CuriousGeorgiy commented on src/box/vy_scheduler.c:
```diff
@@ -1116,6 +1117,140 @@ vy_task_write_run(struct vy_task *task, bool no_compression)
+		W_u = 0;
+	} else {
+		t_r = (double)scnd_last_slice_stmts_count /
+		      last_slice_stmts_count;
```
Why isn't `t_r >= 1` possible like in the case you described above?

EmirVildanov replied:
1.) If `scnd_last_slice_stmts_count` == `new_run_stmts_count`, this case is handled with returning default `bloom_fpr` value;
2.) If `scnd_last_slice_stmts_count` != `new_run_stmts_count`, situation when after the dump there exists a run **X**, which size is bigger than last run size, seems impossible: `vy_range_update_compaction_priority` must have called major compaction after **X** was dumped.

EmirVildanov replied:
Decided to add check that will make sure that none of existing runs are bigger than the last run.
Done.

CuriousGeorgiy commented on changelogs/unreleased/gh-3969-set-bloom-fpr-automatically.md:
```diff
@@ -0,0 +1,3 @@
+## feature/vinyl
+
+* Made bloom filters calculation optimized by run's level in LSM-tree (gh-3969).
```
Please mention the new parameter.

EmirVildanov replied:
Done.

CuriousGeorgiy commented on src/box/index_def.c:
```diff
@@ -48,7 +48,7 @@ const struct index_opts index_opts_default = {
 	/* .run_size_ratio      = */ 3.5,
 	/* .bloom_fpr           = */ 0.05,
-    /* .lookup_cost_coeff   = */ 0.5,
+	/* .lookup_cost_coeff   = */ 0.5,
```
Look's like this should be in e9332b7.

EmirVildanov replied:
Moved it.

CuriousGeorgiy commented on src/box/vy_scheduler.c:
```diff
@@ -1122,7 +1122,7 @@ vy_task_write_run(struct vy_task *task, bool no_compression)
  * W_u - number of unfiltered runs;
  * W_f - number of filtered runs;
- * t_r - Adopted run_size_ratio_coeff.
+ * t_r - Inversed adopted run_size_ratio_coeff (1 / adopted run_size_ratio_coeff).
```
Looks this should be in e9332b7.

EmirVildanov replied:
Moved it.

CuriousGeorgiy commented on perf/bloom_filter/launch.lua:
```diff
@@ -0,0 +1,111 @@
+            removalString = removalString .. " " .. tostring(counter)
+            counter = counter + 1
+        end
+    end
```
Please lookup space id via `box.space._space`.

CuriousGeorgiy commented on perf/bloom_filter/launch.lua:
```diff
@@ -0,0 +1,111 @@
+
+    for _ in pairs(spaces_lookup_cost_coeffs_pairs) do
+        for _ in pairs(bench_run_tuple_counts) do
+            removalString = removalString .. " " .. tostring(counter)
```
Looks questionable.

EmirVildanov replied:
Because `removalString` is calculated with every call of `removeSpaces` or why?

EmirVildanov replied:
Resolved during the call. Decided to leave it for proper cleaning.

CuriousGeorgiy commented on perf/bloom_filter/launch.lua:
```diff
@@ -0,0 +1,111 @@
+local function cleanup()
+    removeSpaces()
+    bench_module_unload()
+    schema_drop()
```
The order looks broken, shouldn't `removeSpaces` be in the end?

EmirVildanov replied:
Changed the order.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+};
+
+/** Trivial wrapper for storing benchmark execution time. */
+struct BenchTimeResult {
```
Looks like overengineering.

EmirVildanov replied:
Removed the wrapper.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+{
+	uint32_t arg_count = mp_decode_array(&args);
+	uint32_t name_len = 0;
+	const char *name = mp_decode_str(&args, &name_len);
```
Looks like this variable is not used, please remove it.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+int
+init(box_function_ctx_t *ctx, const char *args, const char *args_end)
+{
+	uint32_t arg_count = mp_decode_array(&args);
```
Please add an assertion that the argument count is correct.

EmirVildanov replied:
Done.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+
+	memset(tuple_string, '.', STRING_FIELD_SIZE);
+	nonexistenetTuple = new Tuple(tuple_count + 1);
+	data = new TupleData(tuple_count);
```
Please use unique pointers.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+
+	memset(tuple_string, '.', STRING_FIELD_SIZE);
+	nonexistenetTuple = new Tuple(tuple_count + 1);
+	data = new TupleData(tuple_count);
```
Please use unique pointers.

EmirVildanov replied:
How to use it in pair with C++11?

EmirVildanov replied:
Raised version of CMake to 20. Done.

CuriousGeorgiy commented on src/box/box.cc:
```diff
@@ -89,6 +89,7 @@
 #include "mp_uuid.h"
```
> vinyl: added box API
> function for getting bloom
> size memory used by space
>
> NO_CHANGELOG=internal
> NO_DOC=internal
> NO_TEST=internal

Used by index? Please format your sentence and commit subject properly. The right subsystem for this change should be box.

EmirVildanov replied:
Am I right that you mean _net.box_? I don't see _box_ subsystem in the [guide](https://www.tarantool.io/en/doc/latest/dev_guide/developer_guidelines/#how-to-write-a-commit-message).

EmirVildanov replied:
Created an updating issue. Changed to _box_ subsystem. Done.

CuriousGeorgiy commented on extra/exports:
```diff
@@ -12,6 +12,7 @@
 base64_bufsize
 base64_decode
 base64_encode
+box_index_bloom_bsize
```
This should be in b68c98e.

EmirVildanov replied:
Moved it.

CuriousGeorgiy commented on perf/bloom_filter/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+add_library(bench SHARED bloom_filter.cc)
+# for xmalloc
+target_link_libraries(bench unit)
```
Why do you link with `unit`?

EmirVildanov replied:
Because I thought xmalloc is implemented in `unit`. Isn't it true?

EmirVildanov replied:
Deleted this line. Compiling without it.

CuriousGeorgiy commented on extra/exports:
```diff
@@ -12,6 +12,7 @@
 base64_bufsize
```
> Co-authored-by: Nikita
> Pettik
> <korablev@tarantool.org>

Looks like your editor is broken, this should be on a single line.

EmirVildanov replied:
Fixed.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+/** File to save benchmark results. */
+static std::ofstream resultsFile;
+/** File name. */
+static std::string resultsFileName = "results.yml";
```
This can throw, please use `const char []` or `string_view`.

CuriousGeorgiy replied:
You can use a string view literal here and then pass it to `open` via `string_view.data`.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+static constexpr std::uint32_t
+TUPLE_FIELD_COUNT = 2;
+/** Tuple string for filling space. */
+static char tuple_string[STRING_FIELD_SIZE];
```
Please use `std::array`.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+	std::cout << "-----------------------------------------------------------------" << std::endl;
+	std::cout << "STARTED benchmarking." << std::endl;
+
+	memset(tuple_string, '.', STRING_FIELD_SIZE);
```
Please use `std::fill`.

EmirVildanov replied:
Done.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+			  << " does exist. Starting to write." << std::endl;
+		resultsFile.close();
+		resultsFile.open(resultsFileName, std::ios_base::app);
+	}
```
This code looks redundant, why not use `std::ios_base::app` from the beginning?

EmirVildanov replied:
If I understand it right, `std::ios_base::app`would always think that file exists, but we need to add _builds:_ line in case it doesn't.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+		    << result.space_name << std::endl;
+	resultsFile << "                            insert: "
+		    << 1000 * tuple_count / result.insertTimeMs.elapsed_time << std::endl;
+	resultsFile << "                            select: "
```
Please replace all notions of `select` with `get`, it's a little confusing.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+		if (res == -1) {
+			std::cout << "Error occurred while executing select query." << std::endl;
+			exit(1);
+		}
```
Please replace this with an assertion, this can actually slow down your benchmark. And add a similar check for `insert`.

EmirVildanov replied:
Done (commented error reporting variant)

EmirVildanov replied:
Get the answer.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+
+	delete data;
+	delete nonexistenetTuple;
+	resultsFile.close();
```
This function is called by the destructor of `std::ofstream`.

EmirVildanov replied:
Removed this line.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+
+	BenchArgs(BenchArgs &&other) = delete;
+
+	~BenchArgs() { free((void *)space_name); }
```
Looks awful, please use `std::string`.

EmirVildanov replied:
Done.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+
+	~TupleData() {
+		for (auto & t : selectTuples)
+			free(t.begin);
```
Looks awful.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+
+/** Represents one tuple. */
+struct Tuple {
+	char *begin;
```
Please use `std::array<uint8_t, >` here: this way you won't have to mess with allocation.

EmirVildanov replied:
I will try to find an example of using array for buffer.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+	 */
+	explicit Tuple(std::uint32_t pk_value)
+	{
+		std::size_t tuple_size =
```
Please use `std::uint32_t` here, since all `msgpuck_sizeof_*` functions return this type.

EmirVildanov replied:
Done.

CuriousGeorgiy commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+	char *end;
+
+	char *index_begin;
+	char *index_end;
```
Please rename it to `key_*`.

EmirVildanov replied:
Done.

EmirVildanov commented on src/box/box.h:
```diff
@@ -558,6 +558,19 @@ box_sequence_reset(uint32_t seq_id);
+ * \retval 0 on success
+ */
+API_EXPORT int
+box_index_bloom_bsize(uint32_t space_id, uint32_t index_id, size_t *result);
```
Done.

EmirVildanov commented on src/box/box.cc:
```diff
@@ -2834,6 +2835,29 @@ box_session_push(const char *data, const char *data_end)
 }
 
+API_EXPORT int
+box_index_bloom_bsize(uint32_t space_id, uint32_t index_id, size_t *result)
```
Done.

EmirVildanov commented on src/box/box.cc:
```diff
@@ -2834,6 +2835,29 @@ box_session_push(const char *data, const char *data_end)
+		return -1;
+	if (access_check_space(space, PRIV_R) != 0)
+		return -1;
+	index = index_find(space, index_id);
```
Done.

EmirVildanov commented on test/vinyl/bloom.result:
```diff
@@ -45,7 +45,7 @@ box.cfg{vinyl_cache = 0}
 ---
 ...
-_ = s:create_index('pk', {parts = {1, 'unsigned', 2, 'unsigned', 3, 'unsigned', 4, 'unsigned'}})
+_ = s:create_index('pk', {lookup_cost_coeff = 0.05, parts = {1, 'unsigned', 2, 'unsigned', 3, 'unsigned', 4, 'unsigned'}})
```
As soon as in bloom test only one run file is created, the default `lookup_cost_coeff` value for all tests (calculating size and hit/miss statistics) passing is the default value of `bloom_fpr`.

EmirVildanov commented on test/vinyl/bloom.test.lua:
```diff
@@ -18,7 +18,7 @@ s:drop()
 
 s = box.schema.space.create('test', {engine = 'vinyl'})
-_ = s:create_index('pk', {parts = {1, 'unsigned', 2, 'unsigned', 3, 'unsigned', 4, 'unsigned'}})
+_ = s:create_index('pk', {lookup_cost_coeff = 0.05, parts = {1, 'unsigned', 2, 'unsigned', 3, 'unsigned', 4, 'unsigned'}})
```
See above^^^.

EmirVildanov commented on test/vinyl-luatest/gh_3969_bloom_fpr_optimization_test.lua:
```diff
@@ -0,0 +1,160 @@
+    g.server = server:new{alias = 'master'}
+    g.server:start()
+
+    g.update_common_bloom_size_value = function()
```
I have troubles with using `_G` global table: somewhy after inserting big number of tuples in `server:exec()` block, calls of `index:stat().disk.bloom_size` and `index:stat().disk.dump.count` always return 0.

EmirVildanov replied:
Refactoring. See above^^^.

EmirVildanov commented on perf/bloom_filter/bloom_filter.cc:
```diff
@@ -0,0 +1,297 @@
+	uint32_t arg_count = mp_decode_array(&args);
+	uint32_t name_len = 0;
+	const char *name = mp_decode_str(&args, &name_len);
+	const char *build_name = strndup(name, name_len);
```
Done.

EmirVildanov commented on test/vinyl-luatest/common.lua:
```diff
@@ -1,10 +1,12 @@
 local common = {}
```
Done.

EmirVildanov commented on test/vinyl-luatest/gh_3969_bloom_fpr_optimization_test.lua:
```diff
@@ -0,0 +1,160 @@
+local g = t.group()
+
+--
+-- Difficulty: inside of a server:exec block you may call only local things.
```
Currently rewriting tests. See comment below about `_G`.
